### PR TITLE
[Fix #3723] Add new Rails/EnumUniqueness cop.

### LINKED ISSUE #3723: Cop Suggestion: Duplicate "enum" explicit value in Rails
An [enum in Rails](http://edgeapi.rubyonrails.org/classes/ActiveRecord/Enum.html) can be defined either as an array or a hash with specific integer keys defined.

```ruby
# array syntax
enum status: [ :active, :archived ]

# hash syntax
enum status: { active: 0, archived: 1 }
```

With the latter syntax, it's possible to accidentally specify duplicate keys:

```ruby
enum status: { active: 4, archived: 4 }
```

Bugs resulting from this can be hard to identify, and I cannot see a valid use-case for defining duplicate keys in this way.

Is this a reasonable candidate for inclusion as a Rails-specific cop in RuboCop?

Note: Enums are available [since Rails 4.1.0](http://weblog.rubyonrails.org/2014/4/8/Rails-4-1/).

### PR DESCRIPTION
**Pull request by olliebennett**
A new `Rails/EnumUniqueness` cop which identifies any (presumably accidental) usage of duplicate `enum` value definitions.

```ruby
# bad
enum status: { active: 4, archived: 4 }
```

More details in issue #3723.

This is functional (and tested), but probably not the best approach. I'm open to feedback around any shortcuts / simplifications / alternative approaches / other cops to take inspiration from - specifically around my (somewhat clunky) handling of args passed to `enum` via `RuboCop::Node`. I'm also open to alternative names for this cop.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [ ] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

File: spec/rubocop/cop/rails/enum_uniqueness_spec.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  subject(:cop) { described_class.new }
+
+  context 'when array syntax is used' do
+    it 'does not register an offense' do
+      inspect_source(cop, 'enum status: [ :active, :archived ]')
+
+      expect(cop.messages).to be_empty
+    end
+
+    it 'does not crash given additional enum configuration' do
```
Comment (bbatsov) [inline spec/rubocop/cop/rails/enum_uniqueness_spec.rb]: I'd reword this to "does not register an offense when given...". It's a bit odd to have tests testing that something doesn't crash IMO. :-)

**Comment by bbatsov** [comment]
Apart from my small remarks - this has to be rebased.

**Comment by olliebennett** [comment]
Thanks Bozhidar; you're the human Rubocop I can only strive to be! 😉

I've addressed all remarks and rebased again. Thanks for reviewing.

**Comment by bbatsov** [comment]
Btw, I'm sure if this happens or not, but I'm guessing another problem we can check for is duplicate elements in an array enum - e.g. `enum status: [:ala, :bala, :ala]`.

**Comment by olliebennett** [comment]
Good point. Rails at least catches that (albeit indirectly and confusingly) as follows:

> /Users/ollie/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/enum.rb:187:in `detect_enum_conflict!': You tried to define an enum named "status" on the model "User", but this will generate a instance method "ala?", which is already defined by another enum. (ArgumentError)

It would make sense to catch at Rubocop stage though. Will enhance this cop when I get a moment.

**Comment by olliebennett** [comment]
^ I got a moment.
